### PR TITLE
HTTP Request Throttling 

### PR DESCRIPTION
**Pull request by Madhu-1**
### What does this PR achieve? Why do we need it?

This PR adds an HTTP throttling capability to Heketi server.

Heketi is depended on the back-end Gluster storage to perform storage operations, sometimes due to network/Server overloading issue Heketi may not be able to serve n number of requests at a time.
adding an HTTP throttling to limit the number of HTTP POST/DELETE request based on the configuration
which will avoid overload on Heketi and Gluster .once the currently pending requests are completed Heketi will take more POST/DELETE request as usual.

### CODE REVIEW

**Comment by Madhu-1** [comment]
Do not Merge

**Comment by Madhu-1** [comment]
This PR is depending on  #1128  ,

**Comment by Madhu-1** [comment]
@phlogistonjohn  Currently HTTP throttling is not enabled by default in heketi, We need to change the configuration to enable HTTP throttling while running the heketi. The logic behind the HTTP throttling here is, Heketi will serve a limited number of requests at a time (the number is determined by configuration ). Once these requests are completed with success/failure status, heketi is ready to take more requests for processing. By this, we can avoid the overload on the heketi server. we have written a middle-ware which does this job. In #1128 we are supposed to add request ID for X-Request-Id Header for each accepted request for processing. In the throttling middle-ware, we are storing this X-Request-Id as a unique ID as a pending operation in heketi.

 From  HTTP GET requests we extract the request id from the URL.If Heketi has completed the async task associated with this request id, X-Pending Header won't be true anymore, in middle-ware if we get HTTP Success /InternalServerError status code, we are assuming that heketi have completed the task associated with this request id and middle-ware deletes the request id and its corresponding information  from  its in-memory database.

Once the pending request reaches its maximum count in the throttling middle-ware, Heketi starts rejecting further requests for processing, till the pending request is served or throttling cleanup is done.

Suppose if we won't get HTTP GET request corresponding to the X-Request-Id(unique ID)  within a certain interval of time, we assume that task associated with this ID has completed and we delete it from the in-memory and heketi is ready to take more POST/DELETE requests.

File: middleware/http_throttling.go [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+	r.lock.RLock()
+	defer r.lock.RUnlock()
+	return r.Maxcount >= r.ServingCount
+}
+
+//Function to add request id to the queue
+func (r *ReqLimiter) addRequest(reqid string) {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+	r.RequestCache[reqid] = time.Now()
```
[the remark this answers is not in the document]
Comment (Madhu-1) [reply middleware/http_throttling.go]: yeah, it helps to write more test cases.

addressed this changes in new commit

[the remark this answers is not in the document]
Comment (Madhu-1) [reply middleware/http_throttling.go]: yeah, it helps to write more test cases.

addressed this changes in new commit

File: middleware/http_throttling.go [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+	defer r.lock.Unlock()
+	for {
+		select {
+		case <-t.C:
+			for reqID, value := range r.RequestCache {
+				if value.Sub(time.Now()) > (20 * time.Minute) {
+					delete(r.RequestCache, reqID)
+				}
+			}
+		}
```
[the remark this answers is not in the document]
Comment (Madhu-1) [reply middleware/http_throttling.go]: Addressed this comment in the new commit

**Comment by Madhu-1** [comment]
@obnoxxx @phlogistonjohn @raghavendra-talur please review.

**Comment by phlogistonjohn** [comment]
I think there are some ideas here worth considering but if we were to do that I would start with a fresh pr based on master.

**Comment by phlogistonjohn** [comment]
@Madhu-1 if we don't hear any objections in the next 24hrs or so I think I will just close this.

**Pull request closed without merging**